eds semantic html without a separate selector

deepprakash345 opened the pull request:
…tainerEds].....

<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->

## Related Issue

<!--- Every pull request must have a reference to a GitHub or Adobe internal issue. -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes and the overall coverage did not decrease.
- [ ] All unit tests pass on CircleCi.
- [ ] I ran all tests locally and they pass.

jalagari commented on bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java:
```diff
@@ -363,4 +363,20 @@ default String getParentPagePath() {
+    @Nullable
+    @JsonIgnore
+    default Boolean isEdgeDeliveryRequest() {
+        return null;
```
Better to use boolean primitive and return false by default

deepprakash345 commented on ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/page/v1/page/body.html:
```diff
@@ -0,0 +1,21 @@
+<main>
+    <sly data-sly-use.templatedContainer="com.day.cq.wcm.foundation.TemplatedContainer"
+         data-sly-repeat.child="${templatedContainer.structureResources}"
+         data-sly-resource="${child.path @ resourceType=child.resourceType, decorationTagName='div'}"></sly>
```
Hi @vdua , I had this discussion with @jalagari  and @rismehta . we are wrapping everything inside body tag with main tag. We recommend writing styles on the basis of BEM selectors. until someone is searching for body>div, it won't cause a problem.

nit23uec replied:
DOM structure is generally part of public contract and changing DOM hierarchy is backward incompatible.we can't stop anyone from using this dom hierarchy in the code.

rismehta replied:
>DOM structure is generally part of public contract and changing DOM hierarchy is backward incompatible.we can't stop anyone from using this dom hierarchy in the code.

+1, also, we can't expect all customers to consistently adhere to best practices, so this change is not backward compatible.

jalagari replied:
@nit23uec - For transforming HTML to MD main tag is required according to helix API

1. Dont we ask customer to use BEM selectors for styling? 
2. Can you please please the public contract created for DOM hierarchy? Does sites also does it?

As highlighted it will only break for customer who are direct child  selector i.e. body > div

rismehta replied:
> Dont we ask customer to use BEM selectors for styling?

Yes, that's the suggested method, although one could also utilize HTML tags for styling.

> Can you please please the public contract created for DOM hierarchy? Does sites also does it?

Yes both form and sites enforce the same. You can check the component documentation of each component. For example, https://github.com/adobe/aem-core-forms-components/tree/master/ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/textinput/v1/textinput

https://github.com/adobe/aem-core-wcm-components/tree/main/content/src/content/jcr_root/apps/core/wcm/components/button/v2/button

jalagari replied:
@deepprakash345 - lets go with new version of page component as suggested by @rismehta

deepprakash345 replied:
done

nit23uec commented on ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html:
```diff
@@ -13,9 +13,19 @@
-      data-sly-use.grid="com.day.cq.wcm.foundation.model.responsivegrid.ResponsiveGrid"
+<form data-sly-use.grid="com.day.cq.wcm.foundation.model.responsivegrid.ResponsiveGrid"
       data-sly-use.templates="core/wcm/components/commons/v1/templates.html"
       action="${container.metadata.action}"
```
How is container variable defined here?

nit23uec commented on ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/page/v1/page/body.html:
```diff
@@ -0,0 +1,21 @@
+<!--/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  ~ Copyright 2021 Adobe
```
Copyright needs an update

nit23uec commented on bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormContainer.java:
```diff
@@ -363,4 +363,20 @@ default String getParentPagePath() {
+    @Nullable
+    @JsonIgnore
+    default Boolean isEdgeDeliveryRequest() {
+        return null;
```
Request is just a bin servlet request that can originate from anywhere. I'd prefer if we can take this logi. Inside sightly rather than introducing a public method. There is no way to distinguish requests apart of origin and referrer headers of we want to distinguish them based on their origination.

deepprakash345 replied:
Hi nitin, 
Inside sightly , it was not possible to fetch the request attributes. That's why i had to expose a getter api and use it from sightly.

rismehta replied:
> Request is just a bin servlet request that can originate from anywhere. I'd prefer if we can take this logi. Inside sightly rather than introducing a public method. There is no way to distinguish requests apart of origin and referrer headers if we want to distinguish them
based on their origination.

+1, can we take this out of form container and make it part of utils. Form Container should not check such things on where the request originated from. Considering the SRP principle, I would suggest you to do it outside FormContainer

nit23uec replied:
you can make use of `use` API to extract request attribute. we can have a separate `eds.js` in container for the eds use cases.

nit23uec commented on bundles/af-core/src/test/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImplTest.java:
```diff
@@ -230,6 +231,31 @@ void testGetThankYouMessage() throws Exception {
     }
 
+    @Test
+    void testGetIsEdgeDeliveryRequest() throws Exception {
```
The tests seem incomplete for this feature.
we should also test if the request attribute has the property set, then the markup generated should contain pre-code tags.

nit23uec reviewed (changes requested):
inline comments.

nit23uec commented on examples/ui.apps/src/main/content/jcr_root/apps/forms-components-examples/components/form/container/.content.xml:
```diff
@@ -4,4 +4,5 @@
     jcr:title="Adaptive Form Container"
     jcr:description="Add an Adaptive Form to a page."
     sling:resourceSuperType="core/fd/components/form/container/v2/container"
+    cq:styleElements="[div,section,article,main,aside,header,footer]"
```
why do we need `cq:styleElements` for this particular change?

jalagari replied:
@deepprakash345 - Since we are changing main tag in page, we can ignore this changes.

@nit23uec - style element is good have capabilities on container as it provides flexibility  to change container tag type without any new component. Site included in there example

nit23uec replied:
but this is the form container. why would anyone want to have a different tag on the form container other than "form"?

deepprakash345 replied:
done

nit23uec commented on ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/container/v2/container/container.html:
```diff
@@ -13,9 +13,19 @@
 <sly data-sly-test="${wcmmode.edit}" data-sly-use.allowed="com.day.cq.wcm.foundation.AllowedComponentList"></sly>
-<form data-sly-use.container="com.adobe.cq.forms.core.components.models.form.FormContainer"
-      data-sly-use.grid="com.day.cq.wcm.foundation.model.responsivegrid.ResponsiveGrid"
+<form data-sly-use.grid="com.day.cq.wcm.foundation.model.responsivegrid.ResponsiveGrid"
```
now we'll be having a wrapper <div class = "form"> for core components based rendition also, correct?

deepprakash345 replied:
no @nit23uec , the body tag will have a child tag of main which will enclose the entire content of body.
<img width="1728" alt="Screenshot 2024-04-22 at 10 01 48 AM" src="https://github.com/adobe/aem-core-forms-components/assets/16714569/3aff25ad-9508-4ce2-8963-d3069e40b095">

rismehta commented on bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java:
```diff
@@ -359,4 +367,33 @@ public String getName() {
+    @JsonIgnore
+    public String getFormDefinition() {
+        String result = null;
+        try {
```
Can we move this to logic to a utility function

jalagari replied:
Discussed with @rismehta, just move the logic into OperationUtils and reuse it here

deepprakash345 replied:
done

rismehta commented on bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v2/form/FormContainerImpl.java:
```diff
@@ -359,4 +367,33 @@ public String getName() {
+    @Override
+    @Nullable
+    @JsonIgnore
+    public Boolean isEdgeDeliveryRequest() {
```
You don't need to write @JsonIgnore here, if it is already part of interface

jalagari replied:
@deepprakash345 - Discused with @rismehta move utility and use it in HTML similar to parser code

deepprakash345 replied:
done

rismehta reviewed (commented):
check comments

deepprakash345 commented:
@jalagari , @rismehta , @vdua 
made the suggested changes
1. created v2 for page.
2. moved the api **getFormDefinition** to FormStructureParser.
3. moved the api **isEdgeDeliveryRequest** to a newly created file named FormUtil.
4. added the test cases.

vdua commented on bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormUtil.java:
```diff
@@ -0,0 +1,23 @@
+
+@ProviderType
+public interface FormUtil {
+    Boolean isEdgeDeliveryRequest();
```
Do we want multiple implementations of this interface ?
We can directly implement the service without defining the interface.

vdua replied:
Also isn't there any existing util class ?

deepprakash345 replied:
Hi @vdua , We have component Utils but this method doesn't fit in it. This class can be used to put in any misc apis related to form which can't be accomodated in other Utils.

We don't want multiple implementations but coding on top of an interface will be a better coding precedent.

rismehta replied:
>Do we want multiple implementations of this interface ?

This is a provider type as described in the interface, hence only one implementation can exist.

>We can directly implement the service without defining the interface.

Having an interface defines a contract, so even if there is one implementation, there is no harm having an interface

>Also isn't there any existing util class ?

NO

rismehta commented on bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormUtil.java:
```diff
@@ -0,0 +1,23 @@
+import org.osgi.annotation.versioning.ProviderType;
+
+@ProviderType
+public interface FormUtil {
```
Let's rename this to HTLUtils. Form is already there in the package name, hence form in the class name is redundant

deepprakash345 replied:
I feel FormUtils will allow us to write more wider utility methods whereas renaming it to HTLUtils will require creating new utility models if its not required in HTL. Although i can't think of a reason of creating an Util as a Sling model if its not required in HTL.

So renaming this file to HtlUtil

rismehta commented on ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/page/v2/.content.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Page component should have a read me like other components. The readme should describe the reasoning behind this new version

deepprakash345 replied:
done

rismehta commented on bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/FormStructureParser.java:
```diff
@@ -43,4 +43,10 @@ public interface FormStructureParser {
+     *
+     * @return form definition json in Publish view
+     */
+    String getFormDefinition();
```
Please bump the minor version and add a @since tag in javadoc mentioning the new version. You can check the existing code for reference

rismehta reviewed (changes requested):
check comments

rismehta commented on bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/HtlUtil.java:
```diff
@@ -0,0 +1,23 @@
+
+@ProviderType
+public interface HtlUtil {
+    Boolean isEdgeDeliveryRequest();
```
Please add java doc for all API

deepprakash345 replied:
done

rismehta commented on bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/form/HtlUtilImpl.java:
```diff
@@ -0,0 +1,46 @@
+    private SlingHttpServletRequest request;
+
+    public Boolean isEdgeDeliveryRequest() {
+        if (request != null) {
```
`if (request != null) {
    Object isEdgeDelivery = request.getAttribute("com.adobe.aem.wcm.franklin.internal.servlets.FranklinDeliveryServlet");
    return isEdgeDelivery != null && isEdgeDelivery.equals(Boolean.TRUE);
}
return false;`

rismehta commented on ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/page/v2/page/body.html:
```diff
@@ -0,0 +1,21 @@
+<!--/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Can we create a test case of a form using this page component, and check if it renders fine ? This is currently missing in this PR

deepprakash345 replied:
added test cases

rismehta reviewed (changes requested):
Please add test case using the new v2 page component to check as per EDS use-case, you can name the test case as pagev2.runtime.spec.js

rismehta commented on ui.tests/test-module/specs/page/page.runtime.spec.js:
```diff
@@ -0,0 +1,32 @@
+ * limitations under the License.
+ ******************************************************************************/
+
+describe("Form with Number Input", () => {
```
Please rename the test describe correctly

deepprakash345 replied:
done

rismehta commented on ui.tests/test-module/specs/page/page.runtime.spec.js:
```diff
@@ -0,0 +1,32 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
```
Please change the spec to pagev2.runtime.spec.js

deepprakash345 replied:
done

rismehta commented on ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/page/v1/page/README.md:
```diff
@@ -0,0 +1,30 @@
+-->
+Adaptive Form Core Page (v1)
+====
+Adaptive Form Core Page component v1 which extends the wcm v3 [Page](https://github.com/adobe/aem-core-wcm-components/tree/d50ba1cb0d6544870daff71281bf9dfd3f6fbeb0/content/src/content/jcr_root/apps/core/wcm/components/page/v3/page) Component.
```
The reference does not point to master branch, please change link

deepprakash345 replied:
main branch link added

rismehta commented on ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/page/v1/page/README.md:
```diff
@@ -0,0 +1,30 @@
+-->
+Adaptive Form Core Page (v1)
+====
+Adaptive Form Core Page component v1 which extends the wcm v3 [Page](https://github.com/adobe/aem-core-wcm-components/tree/d50ba1cb0d6544870daff71281bf9dfd3f6fbeb0/content/src/content/jcr_root/apps/core/wcm/components/page/v3/page) Component.
```
you don't need to mention the specific version of WCM core component

deepprakash345 replied:
version removed

rismehta reviewed (changes requested):
check comments
